### ISSUE TITLE
Terminal Toggling 

### ISSUE THREAD

**Issue by jasmcaus** [NvChad/NvChad#489]
**Describe the bug**
As per [this](https://github.com/NvChad/NvChad/blob/main/lua/core/default_config.lua#L133) (see code snippet below):
```lua
   -- terminal related mappings
   terminal = {
      -- multiple mappings can be given for esc_termmode and esc_hide_termmode
      -- get out of terminal mode
      esc_termmode = { "jk" }, -- multiple mappings allowed
      -- get out of terminal mode and hide it
      esc_hide_termmode = { "JK" }, -- multiple mappings allowed
      -- show & recover hidden terminal buffers in a telescope picker
      pick_term = "<leader>W",
      -- below three are for spawning terminals
      new_horizontal = "<leader>h",
      new_vertical = "<leader>v",
      new_window = "<leader>w",
   },
```
I'm able to spawn a new terminal (using `<leader>h`) and hide it using `JK`, but I'm unable to reopen the same terminal. Using `<leader>h` spawns a brand new terminal. Now, I know I can use `<leader>W` to enter Telescope view and select a terminal, but that's not ideal. Is there a shortcut to reopen a previous terminal buffer?

**Comment by G-Rowell** [comment]
Ooohhh,

I'll have a read through this in a few minutes when I get home.
Terminals are my area in NvChad so I'll do what I can!

G

**Comment by G-Rowell** [comment]
This is a great question!

Currently we don't have a bind to re-open the last used terminal buffer.
I think that's a good idea though, and not too hard to implement.

---

What are your thoughts on:

`<leader> + H` to re-open the last used horizontal term
`<leader> + V` to re-open the last used vertical term

**Comment by jasmcaus** [comment]
Seems okay. I was approaching it more from the side of VSCode where `<C-j>` spawns a new terminal (if none exist), and subsequent keystrokes (of `<C-j>`) repeatedly enable and hide the terminal, with no new terminals getting spawned each time those keys are pressed. 

I imagine that this would be difficult to implement for Vim, but here are my two cents:
* `<leader> + H`to spawn a new terminal
* `<leader> + J` to enable an already existing terminal
* `<leader> + K` to hide a terminal

Of course, if you have multiple terminal buffers, the already existing `<leader> + W` seems to work just as fine.

**Comment by Daniel-Boll** [comment]
I also love the workflow that I use to had on *tmux* before using the integrated terminal where I open a vertical terminal alongside and then if I want to focus on my code I switch to the code pane and `<leader> + z` to sort of zoom in, then to go back to the other pane I could `<leader> + z` again an then switch with `alt + l` or straight up using `alt + l` that would produce the same result.

Not sure if something like this is possible or even wanted.

**Comment by jasmcaus** [comment]
@Daniel-Boll right, so pressing, say, `<leader> + H` would spawn a new terminal (either horizontal/vertical depending on your preference). As for `alt + l`/`<leader> + z`, I'm not too familiar with tmux, so wouldn't be able to say much about it.

**Comment by G-Rowell** [comment]
Implementing this functionality would for sure be possible!

We're currently sorting out some administrative things regarding NvChad though,
so hold on a bit but I do like this idea!

**Comment by zbirenbaum** [comment]
I like the toggleterm plugin and its pretty easy to configure. If you want a fast solution I would recommend looking into it, you can set mappings and dimensions for horizontal/vertical, and select specific terminals by their assigned number, show and hide a terminal with same hotkey, and move between termbuffer and other buffers with something like <C-h/j/k/l>. I believe the git page for it specifically includes the 'jk' to escape terminal mapping. I can provide my config files as well if you'd like.

https://github.com/akinsho/toggleterm.nvim

**Comment by tydra-wang** [comment]
btw, `<leader>x` just hide the terminal window but not delete the buffer.

**Comment by siduck** [comment]
> btw, `<leader>x` just hide the terminal window but not delete the buffer.

yes it will hide a terminal , if there's a buffer then it will close that buffer. 

if you want to close that terminal buffer , try :bd! , this works so just add a new mapping for it.

**Comment by Daniel-Boll** [comment]
The toggle workflow provided by the [numToStr/FTerm](https://github.com/numToStr/FTerm.nvim) is just lovely, if we could achieve something similar would be awesome.

**Comment by siduck** [comment]
> The toggle workflow provided by the [numToStr/FTerm](https://github.com/numToStr/FTerm.nvim) is just lovely, if we could achieve something similar would be awesome.

hmm what does it do? toggling floating terminal?

**Comment by zbirenbaum** [comment]
> > The toggle workflow provided by the [numToStr/FTerm](https://github.com/numToStr/FTerm.nvim) is just lovely, if we could achieve something similar would be awesome.
> 
> 
> 
> hmm what does it do? toggling floating terminal?

Yeah but I've used FTerm in the past and I'm not sure what triggers it, but issues that crash not just nvim but my whole shell due to segfaults were common.

Toggleterm has a floating terminal functionality built in, I use alt-i to bring up a float term and ctrl-x to bring up a normal term at the bottom. 

Tbh toggle term is so efficient and comes with so many features (like term numbering, auto entry, and escape sequences) out of the box I would consider putting it in core rather than reimplementing tons of functionality and expanding the codebase.

**Comment by siduck** [comment]
> > > The toggle workflow provided by the [numToStr/FTerm](https://github.com/numToStr/FTerm.nvim) is just lovely, if we could achieve something similar would be awesome.
> > 
> > 
> > hmm what does it do? toggling floating terminal?
> 
> Yeah but I've used FTerm in the past and I'm not sure what triggers it, but issues that crash not just nvim but my whole shell due to segfaults were common.
> 
> akinsho/toggleterm.nvim has a floating terminal functionality built in, I use alt-i to bring up a float term and ctrl-x to bring up a normal term at the bottom.
> 
> Tbh toggle term is so efficient and comes with so many features (like term numbering, auto entry, and escape sequences) out of the box I would consider putting it in core rather than reimplementing tons of functionality and expanding the codebase.

hmm , the issue with adding toggleterm might be creating more issues for us . 

ex  : https://github.com/akinsho/toggleterm.nvim/issues/111

**Comment by siduck** [comment]
our implementation allows toggling of multiple terminals , previewing them in the telescope too.

**Comment by zbirenbaum** [comment]
I've never had that problem and I use it. Let me grab my config file real quick

**Comment by peppapig450** [comment]
the battle with github actions 😂

**Comment by zbirenbaum** [comment]
I managed to get a minimal implementation of creating a terminal and hiding (but not deleting) it with just the nvim api, but there are some issues with trying to reshow it.  Setting the current window with  nvim_set_current_win(id) and then doing vim.cmd('hide') works fine, but doing vim.cmd('unhide') and then setting the current win to the same id does not work. The terminal will reopen with old command output present if you scroll up after running ls for instance and hiding it, then reshowing it, but it will not autofocus on reshow properly, and ruins the saved IDs list.

**Comment by zbirenbaum** [comment]
@siduck @G-Rowell I have toggling the same terminal on and off working perfectly on my end. Will the telescope stuff mentioned earlier in here work by default with a normal terminal buffer, or do I need to add in some stuff to get the functionality of the current terminal.

**Comment by siduck** [comment]
> @siduck @G-Rowell I have toggling the same terminal on and off working perfectly on my end. Will the telescope stuff mentioned earlier in here work by default with a normal terminal buffer, or do I need to add in some stuff to get the functionality of the current terminal.

oh cool but we cant use toggleterm :c

**Comment by zbirenbaum** [comment]
> > @siduck @G-Rowell I have toggling the same terminal on and off working perfectly on my end. Will the telescope stuff mentioned earlier in here work by default with a normal terminal buffer, or do I need to add in some stuff to get the functionality of the current terminal.
> 
> 
> 
> oh cool but we cant use toggleterm :c 

No I'm saying I implemented toggleterm-esque functionality myself without all the extra stuff that's it does.

**Comment by zbirenbaum** [comment]
@siduck 

Ended up making it more fully featured since it was a good activity to learn about how to manage buffers and windows properly, so it kinda is a full plugin now. The filetype and stuff is the exact same as in NvChad. So everything should work. Just put this in your plugin table:

```
{
      "zbirenbaum/nvim-chadterm",
      setup = function ()
         vim.schedule_wrap(vim.cmd[[packadd nvim-chadterm]])
      end,
      config = function ()
         vim.schedule_wrap(require("chadterm").setup({}))
      end
  },
```

Make sure not to skip out on the schedule wrap stuff. The default hotkeys in it are set to <leader>h and <leader>v which are NvChad's defaults as well if I am not mistaken. The schedule portion ensures that NvChad won't override the hotkeys, as well as lazy loads everything.

You will probably need to comment out the terminal related stuff in autocommands as well.

**Comment by siduck** [comment]
oh :l

**Comment by zbirenbaum** [comment]
@siduck it now works for non-nightly versions as well and will detect users not on .7 and use nvim_set_keymap instead. Go ahead and try it out.

**Comment by siduck** [comment]
> @siduck it now works for non-nightly versions as well and will detect users not on .7 and use nvim_set_keymap instead. Go ahead and try it out.

I thought the terms feature in our extensions plug-in was gonna get fixed , I'll fix it myself someday

**Comment by zbirenbaum** [comment]
> > @siduck it now works for non-nightly versions as well and will detect users not on .7 and use nvim_set_keymap instead. Go ahead and try it out.
> 
> I thought the terms feature in our extensions plug-in was gonna get fixed , I'll fix it myself someday

I thought this was an issue regarding being able to retoggle terminals?

**Comment by zbirenbaum** [comment]
If you want it to be smaller, I can take out most of the customization features and squash it into one commit you can pull to the extensions repo. Toggling the terminal instead of opening a new one is deceptively complicated and it took about 5 hours of just reading documentation to get it right. This isn't something you can do in a one-liner unfortunately.

**Comment by siduck** [comment]
> If you want it to be smaller, I can take out most of the customization features and squash it into one commit you can pull to the extensions repo. Toggling the terminal instead of opening a new one is deceptively complicated and it took about 5 hours of just reading documentation to get it right. This isn't something you can do in a one-liner unfortunately.

Im not talking about it smaller but i wanted to have our extensions have this functionality, if that's possible then you can open a PR

**Comment by zbirenbaum** [comment]
> > If you want it to be smaller, I can take out most of the customization features and squash it into one commit you can pull to the extensions repo. Toggling the terminal instead of opening a new one is deceptively complicated and it took about 5 hours of just reading documentation to get it right. This isn't something you can do in a one-liner unfortunately.
> 
> Im not talking about it smaller but i wanted to have our extensions have this functionality, if that's possible then you can open a PR

I just confirmed it works perfectly fine with the extensions (assuming by that you mean the open in telescope and bufferline stuff) 

Refactoring the code to take options the way they are currently done in chadrc and default config will take a bit (but not too long, mostly done already) however, putting the mappings and their corresponding handlers in mappings.lua will be... difficult to say the least.

**Comment by zbirenbaum** [comment]
@siduck  #811

**Comment by jasmcaus** [comment]
@zbirenbaum didn't have a chance to look at it, but does it allow toggling of the same terminal buffer? For instance, if I spawn a new terminal and `npm` something, and hide it, can I spawn the *same* buffer using a hotkey?

**Comment by zbirenbaum** [comment]
> @zbirenbaum didn't have a chance to look at it, but does it allow toggling of the same terminal buffer? For instance, if I spawn a new terminal and `npm` something, and hide it, can I spawn the _same_ buffer using a hotkey?

Yes, that is the whole purpose of the plugin. 

Unfortunately, while many sizable improvements have been made since I posted that (And I found numerous bugs that were not fixed in the original repo as Siduck wants them in the extensions repository, not external) I still have not had time to complete it. IIRC, the status of the one in #811 is that it now allows selection from telescope from the full buffers list, but did not display properly under the terminal specific selector.

If you do try out that branch and catalog any issues you find, it would definitely go a long way towards helping me complete this sooner.

**Comment by zbirenbaum** [comment]
#811 is fully implemented and testing is encouraged. Any users who wish to try it out can clone zbirenbaum/NvChad and checkout the terminal_toggle branch. Then, go into init.lua and change
```
   {
      "NvChad/extensions",
      config = function ()
         vim.schedule_wrap(require("nvchad.terminal").init())
      end
   },
```
to

```
   {
      "zbirenbaum/extensions",
      config = function ()
         vim.schedule_wrap(require("nvchad.terminal").init())
      end
   },
```
Since @siduck requested the terminal implementation be moved to the extensions repo rather than core NvChad. You will *need* to do this step to test the new features, as the terminal mappings in mappings.lua are changed to use the module from extensions, but the changes there aren't merged yet.

I think this is ready to be merged from my testing, but this is essentially an entire plugin in itself and edge cases are possible, so any input is appreciated.
